src/jdk.incubator.code/share/classes/jdk/incubator/code/internal/CodeModelToAST.java

```diff
@@ -0,0 +1,287 @@
+            default -> {
+                if (!jtToType.containsKey(te)) {
+                    throw new IllegalStateException("JavaType -> Type not found for: " + te.externalize().toString());
+                }
+                yield jtToType.get(te);
+            }
+        };
+    }
+
+    private JCTree invokeOpToJCMethodInvocation(CoreOp.InvokeOp invokeOp) {
+        Method method;
+        try {
+            method = invokeOp.invokeDescriptor().resolveToDirectMethod(MethodHandles.lookup());
+        } catch (ReflectiveOperationException e) {
+            throw new RuntimeException(e);
+        }
+        long flags = method.getModifiers();
+        var name = names.fromString(invokeOp.invokeDescriptor().name());
+        Value receiver = invokeOp.invokeKind() == CoreOp.InvokeOp.InvokeKind.INSTANCE ? invokeOp.operands().get(0) : null;
+        List<Value> arguments = invokeOp.operands().stream().skip(receiver == null ? 0 : 1).collect(List.collector());
+        var paramTypes = invokeOp.invokeDescriptor().type().parameterTypes().stream().map(this::typeElementToType)
+                .collect(List.collector());
+        var restype = typeElementToType(invokeOp.resultType());
+        var type = new Type.MethodType(paramTypes, restype, List.nil(), syms.methodClass);
+        var methodSym = new Symbol.MethodSymbol(flags, name, type,
+                typeElementToType(invokeOp.invokeDescriptor().refType()).tsym);
+        var meth = receiver == null ? treeMaker.Ident(methodSym) : treeMaker.Select((JCTree.JCExpression) valueToTree.get(receiver), methodSym);
+        var args = new ListBuffer<JCTree.JCExpression>();
+        for (Value operand : arguments) {
+            args.add((JCTree.JCExpression) valueToTree.get(operand));
+        }
+        var methodInvocation = treeMaker.App(meth, args.toList());
+        if (invokeOp.isVarArgs()) {
+            var lastParam = invokeOp.invokeDescriptor().type().parameterTypes().getLast();
+            Assert.check(lastParam instanceof ArrayType);
+            methodInvocation.varargsElement = typeElementToType(((ArrayType) lastParam).componentType());
+        }
+        if (invokeOp.result().uses().isEmpty()) {
+            return treeMaker.Exec(methodInvocation);
+        }
+        return methodInvocation;
+    }
+
+    private JCTree opToTree(Op op) {
+        JCTree tree = switch (op) {
+            case CoreOp.ConstantOp constantOp when constantOp.value() == null ->
+                    treeMaker.Literal(TypeTag.BOT, null).setType(syms.botType);
+            case CoreOp.ConstantOp constantOp -> treeMaker.Literal(constantOp.value());
+            case CoreOp.InvokeOp invokeOp -> invokeOpToJCMethodInvocation(invokeOp);
+            case CoreOp.NewOp newOp -> {
+                if (newOp.resultType() instanceof ArrayType at) {
+                    var elemType = treeMaker.Ident(typeElementToType(at.componentType()).tsym);
+                    // @@@ we assume one dimension
```

Comment:
note: `ArrayType` has a `dimensions()` method -- so in principle you could use that and keep fetching dimension operands (and update the type of the array to be built)

Reply:
I updated the code.